module/zfs/spa_config.c

```diff
@@ -368,17 +368,24 @@ spa_write_cachefile(spa_t *target, boolean_t removing, boolean_t postsysevent,
  * information for all pool visible within the zone.
  */
 nvlist_t *
-spa_all_configs(uint64_t *generation)
+spa_all_configs(uint64_t *generation, int *error)
 {
 	nvlist_t *pools;
 	spa_t *spa = NULL;
 
-	if (*generation == spa_config_generation)
+	if (*generation == spa_config_generation) {
+		*error = EEXIST;
 		return (NULL);
+	}
 
 	pools = fnvlist_alloc();
 
-	mutex_enter(&spa_namespace_lock);
+	*error = mutex_enter_interruptible(&spa_namespace_lock);
+	if (*error) {
+		fnvlist_free(pools);
+		*error = EINTR;
```

Comment:
The second assignment to *error is since the first will get a value < 0 but later SET_ERROR(error) wants a value > 0. 
I don't really like this method of returning an error, and I think it would be nicer to use Linux's ERR_PTR / IS_ERR macros to return an error value via the function's return value. But this won't work since that's not portable to FreeBSD. I do think it's important to be able to distinguish the different error cases (which can't be done by just checking for NULL). Any ideas on a better way to do this?

Reply:
Yea it's a little ugly, but I don't see a better way of doing it.  I think they way you have it is fine.

Reply:
Let's flip this around.  Something like this would be a bit more conventional.

```c
int
spa_all_configs(uint64_t *generation, nvlist **pools)
{
        spa_t *spa = NULL;

        if (*generation == spa_config_generation)
                return (SET_ERROR(EEXIST));

        int error = mutex_enter_interruptible(&spa_namespace_lock);
        if (error)
                return (SET_ERROR(EINTR));

        *pools = fnvlist_alloc();
        while ((spa = spa_next(spa)) != NULL) {
                ....
                
                
       return (0);
```

Reply:
Thanks, that is nicer. Updated in the latest revision.